python/perspective/perspective/tests/table/test_table_object.py

```diff
@@ -0,0 +1,204 @@
+# *****************************************************************************
+#
+# Copyright (c) 2019, the Perspective Authors.
+#
+# This file is part of the Perspective library, distributed under the terms of
+# the Apache License 2.0.  The full license can be found in the LICENSE file.
+#
+import six
+from perspective.table import Table
+from datetime import date, datetime
+
+class CustomObjectBlank(object):
+    pass
+
+class CustomObjectStore(object):
+    def __init__(self, value):
+        self._value = value
+
+    def _psp_dtype_(self):
+        return "object"
+
+    def __int__(self):
+        return int(self._value)
+
+    def __repr__(self):
+        return 'test'
+
+class CustomObjectRepr(object):
+    def __init__(self, value):
+        self._value = value
+
+    def __repr__(self):
+        return str(self._value)
+
+class CustomObjectIntPromoteToString(CustomObjectRepr):
+    def _psp_dtype_(self):
+        return int
+
+class CustomObjectFloatPromoteToString(CustomObjectRepr):
+    def _psp_dtype_(self):
+        return float
+
+class CustomObjectIntBoth(CustomObjectRepr):
+    def _psp_dtype_(self):
+        return int
+
+    def _psp_repr_(self):
+        return int(self._value) + 1
+
+class CustomObjectFloatBoth(CustomObjectRepr):
+    def _psp_dtype_(self):
+        return float
+
+    def _psp_repr_(self):
+        return float(self._value) + 1.0
+
+class CustomObjectIntConvert(CustomObjectRepr):
+    def _psp_dtype_(self):
+        return int
+
+    def __int__(self):
+        return int(self._value)
+
+    def __repr__(self):
+        return 'test'
+
+class CustomObjectFloatConvert(CustomObjectRepr):
+    def _psp_dtype_(self):
+        return float
+
+    def __float__(self):
+        return float(self._value)
+
+    def __repr__(self):
+        return 'test'
+
+
+class CustomObjectIntConvertFromFloat(CustomObjectRepr):
+    def _psp_dtype_(self):
+        return int
+
+    def __float__(self):
+        return float(self._value)
+
+    def __repr__(self):
+        return 'test'
+
+class CustomObjectFloatConvertFromInt(CustomObjectRepr):
+    def _psp_dtype_(self):
+        return float
+
+    def __int__(self):
+        return int(self._value)
+
+    def __repr__(self):
+        return 'test'
+
+class TestTableObjectsExtract(object):
```

Comment:
Is there a way to test the "storing an object whose pointer has been cleaned" behaviour you mention in the PR notes?

Reply:
yes, we can explicitly `del` the object but we don't want to support this e.g. if we increment the reference count, python won't delete the object so long as psp has a copy of the pointer. We'll then need to make sure that when the table is destroyed/column destroyed/value destroyed/value overwritten that we decrement the reference count so python can deallocate it. 

`pybind`'s `object` class will do this automatically (and its `handle` class will not), but we're not using either of those internally so we'll need to use the underlying APIs (https://docs.python.org/3.7/c-api/refcounting.html)

Reply:
I'll try to think of a good test around our use of reference counting, we can probably just write a test to make a table and count references (should be previous+1), and then destroy the table (which should subtract 1).
https://docs.python.org/dev/library/sys.html#sys.getrefcount